File: garagemanagement.client/src/app/models/job-card.ts
import { ValidatorFn } from "@angular/forms";

export interface JobCard {
  refNo: string;
  jobCardNo: string;
  regNo: string;
  invoiceNo?: string;
  serviceType: string;
  vehicle: string;
  status: string;
  customerName: string;
  mobileNo: string;
  arrivalDate: string;
  arrivalTime: string;
  insuranceCorporate?: string;
  claimNo?: string;
  estDeliveryDate?: string;
  accidentDate?: string;
}
export interface JobCardDto {
  id: number;
  vehicleData: {
    registrationNo: string;
    odometerIn: string;
    avgKmsPerDay: string;
    vin: string;
    engineNo: string;
    vehicleColor: string;
    fuelType: string;
    serviceType: string;
    serviceAdvisor: string;
    technician: string;
    vendor: string;
  };

  customerInfo: {
    corporate: string;
    customerName: string;
    mobile: string;
    alternateMobile: string;
    email: string;
    deliveryDate: string;
    insuranceCompany: string;
  };

  concerns: {
    text: string;
    active: boolean;
  }[];

  advancePayment: {
    cash: string;
    bankName: string;
    chequeNo: string;
    amount: string;
    date: string;
  };
}

export interface VehicleDetailsUI {
  jobCardNumberForEstimation: string;
  regNo: string;
  jobCardNo: string;

  customerName: string;
  mobile: string;
  email: string;

  odometer: number;
  model: string;
  fuelType: string;
  vin: string;
  engineNo: string;
}

export interface EstimationItem {
  name: string;
  type: string;
  partNo: string;
  rate: number;
  discount: number;
  hsn: string;
  taxPercent: number;
  taxAmount: number;
  total: number;
  approval: string;
  reason: string;
}
export type ColumnType = 'text' | 'number' | 'date' | 'select';
export interface PopupColumnConfig {
  field: string;
  header: string;
  type: ColumnType;
  validators?: ValidatorFn[];
  options?: { label: string; value: any }[];
  getOptions?: (row: any) => { label: string; value: string }[];
  minDate?: string;
  maxDate?: string;
}

export interface PopupTabConfig {
  tabKey: string;
  label: string;
  columns?: PopupColumnConfig[];
  isTextarea?: boolean;
  allowAdd?: boolean;
  allowDelete?: boolean;
}

File: GarageManagement.Server/Model/JobCard.cs
namespace GarageManagement.Server.Model
{
    public class JobCard
    {
        public long Id { get; set; }

        // Vehicle
        public string RegistrationNo { get; set; } = string.Empty;
        public string? JobCardNo { get; set; } = string.Empty;
        public long? OdometerIn { get; set; }
        public long? AvgKmsPerDay { get; set; }
        public string Vin { get; set; } = string.Empty;
        public string EngineNo { get; set; } = string.Empty;
        public string VehicleColor { get; set; } = string.Empty;
        public string FuelType { get; set; } = string.Empty;
        public string ServiceType { get; set; } = string.Empty;
        public string ServiceAdvisor { get; set; } = string.Empty;
        public string Technician { get; set; } = string.Empty;
        public string Vendor { get; set; } = string.Empty;

        // Customer
        public string Corporate { get; set; } = string.Empty;
        public string CustomerName { get; set; } = string.Empty;
        public string Mobile { get; set; } = string.Empty;
        public string AlternateMobile { get; set; } = string.Empty;
        public string Email { get; set; } = string.Empty;
        public DateTime? DeliveryDate { get; set; }
        public string InsuranceCompany { get; set; } = string.Empty;

        // Popup (single column)
        public decimal Discount { get; set; }
        public decimal Paid { get; set; }
        public string? ServiceSuggestions { get; set; } = string.Empty;
        public string? Remarks { get; set; } = string.Empty;
        public string? Status { get; set; } = string.Empty;

        // Navigation
        public JobCardAdvancePayment? AdvancePayment { get; set; }
        public ICollection<JobCardConcern> Concerns { get; set; } = new List<JobCardConcern>();
        public ICollection<JobCardEstimationItem>? JobCardEstimationItems { get; set; }
        public ICollection<JobCardTyreBattery>? TyreBatteries { get; set; }
        public ICollection<JobCardCancelledInvoice>? CancelledInvoices { get; set; }
        public ICollection<JobCardCollection>? Collections { get; set; }

    }
}
